File: C++/4th_ex/main.cpp
#include "common.h"
#include "encypt_decrypt.h"
#include <iostream>

using namespace std;

int main() {
  int64_t p = gen_prime_num();
  int64_t q = gen_prime_num();
  int64_t n = p * q;
  int64_t phi = (p - 1) * (q - 1);
  int64_t e = find_coprime_num(phi);
  int64_t d = (ex_euclid(e, phi) % phi + phi) % phi;

  cout << "public keys: " << e << ' ' << n << endl;
  cout << "closed keys: " << d << ' ' << n << endl;

  string m;
  getline(cin, m);

  int64_t session_key = gen_rand_num(1, n - 1);
  int64_t enc_key = mod_pow(session_key, e, n);

  string enc_message = XOR(m, session_key);
  cout << enc_message << endl;

  int64_t dec_key = mod_pow(enc_key, d, n);
  string dec_message = XOR(enc_message, dec_key);
  cout << dec_message << endl;

  return 0;
}

File: python/4th_ex/main.py
from random import randint
import common
import encrypt_decrypt

if __name__ == '__main__':
    q = common.gen_prime_num()
    p = common.gen_prime_num()
    n = p * q
    phi = (p - 1) * (q - 1)
    e = common.find_coprime_num(phi)
    d = (common.ex_euclid(e, phi) % phi) + phi % phi

    print(f"public keys: {e} {n}")
    print(f"privat keys: {d} {n}")
    
    m = input()
    session_key = randint(1, n - 1)
    enc_key = common.mod_pow(session_key, e, n)
    enc_message = encrypt_decrypt.XOR(m, session_key)
    print(enc_message)

    dec_key = common.mod_pow(enc_key, d, n)
    dec_message = encrypt_decrypt.XOR(enc_message, dec_key)
    print(dec_message)


